tests/TestCase/Database/CommonTableExpressionIntegrationTest.php

```diff
@@ -0,0 +1,942 @@
+                'col' => '1',
+            ],
+        ];
+
+        $stmt = $query->execute();
+        $result = $stmt->fetchAll('assoc');
+        $stmt->closeCursor();
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testWithCteAsCallable(): void
+    {
+        $cteQuery = $this->connection
+            ->newQuery()
+            ->select(['col' => 1]);
+
+        $query = $this->connection
+            ->newQuery()
+            ->with(function (QueryExpression $exp, Query $query) use ($cteQuery) {
+                return $query
+                    ->cte('cte', $cteQuery);
+            })
+            ->select('col')
+            ->from('cte');
+
+        $this->assertQuotedQuery(
+            'WITH cte AS \(SELECT 1 AS <col>\) SELECT <col> FROM <cte>',
+            $query->sql(),
+            !$this->autoQuote
+        );
+
+        $expected = [
+            [
+                'col' => '1',
+            ],
+        ];
+
+        $stmt = $query->execute();
+        $result = $stmt->fetchAll('assoc');
+        $stmt->closeCursor();
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testWithCteInvalidType(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage(
+            'The common table expression must be an instance of `Cake\Database\Expression\CommonTableExpression`, ' .
+            '`integer` given.'
+        );
+
+        $this->connection->newQuery()->with(123);
+    }
+
+    public function testWithCteNamesMustBeUnique(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage(
+            'A common table expression with the name `cte` already exists.'
+        );
+
+        $this->connection
+            ->newQuery()
+            ->with(function (QueryExpression $exp, Query $query) {
+                return $query
+                    ->cte('cte', $query->newExpr('SELECT 1'));
+            })
+            ->with(function (QueryExpression $exp, Query $query) {
+                return $query
+                    ->cte('cte', $query->newExpr('SELECT 1'));
+            });
+    }
+
+    public function testWithRecursiveCte(): void
+    {
+        $query1 = $this->connection
+            ->newQuery()
+            ->select(1);
+        $query2 = $this->connection
+            ->newQuery()
+            ->select(function (Query $query) {
+                return $query->newExpr('col + 1');
+            })
+            ->from('cte')
+            ->where(['col !=' => 3], ['col' => 'integer']);
+        $cteQuery = $query1->unionAll($query2);
+
+        $query = $this->connection
+            ->newQuery()
+            ->with(function (QueryExpression $exp, Query $query) use ($cteQuery) {
+                return $query
+                    ->cte('cte', $cteQuery)
+                    ->setFields(['col'])
+                    ->setRecursive(true);
```

Comment:
Should this method be an enableRecursive/disableRecurisve pair?

Reply:
I was unsure about that, as it doesn't really enable/disable recursion for the CTE itself, but rather flags it at using a recursive query.

Reply:
Yea, enable/disable doesn't work because you don't want to "enable" a recursive cte without it being a recursive cte. Enabling implies you allow CTE's to be recursive instead of flagging them.

Reply:
Ok. setRecursive() works for me.